stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,311 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// implement atomic wait / notify_one / notify_all
+
+// clang-format off
+
+#include <atomic>
+#include <cstdint>
+#include <new>
+#include <thread>
+#include <Windows.h>
+
+// clang-format on
+
+namespace {
+
+    constexpr size_t _Wait_table_size_power = 8;
+    constexpr size_t _Wait_table_size       = 1 << _Wait_table_size_power;
+    constexpr size_t _Wait_table_index_mask = _Wait_table_size - 1;
+
+    struct _Wait_context {
+        const void* _Storage; // Pointer to wait on
+        _Wait_context* _Next;
+        _Wait_context* _Prev;
+        CONDITION_VARIABLE _Condition;
+    };
+
+    struct _Guarded_wait_context : _Wait_context {
+        _Guarded_wait_context(const void* _Storage_, _Wait_context* const _Head) noexcept
+            : _Wait_context{_Storage_, _Head, _Head->_Prev, CONDITION_VARIABLE_INIT} {
+            _Prev->_Next = this;
+            _Next->_Prev = this;
+        }
+
+        ~_Guarded_wait_context() {
+            _Wait_context* const _Next_local = _Next;
+            _Wait_context* const _Prev_local = _Prev;
+            _Next->_Prev                     = _Prev_local;
+            _Prev->_Next                     = _Next_local;
+        }
+
+        _Guarded_wait_context(const _Guarded_wait_context&) = delete;
+        _Guarded_wait_context& operator=(const _Guarded_wait_context&) = delete;
+    };
+
+    class _SrwLock_guard {
+        SRWLOCK* _Locked;

```

Comment:
We conventionally mention `private` explicitly for classes, even though that's the default:
```suggestion
    class _SrwLock_guard {
    private:
        SRWLOCK* _Locked;
```